firedrake/mg/ufl_utils.py

```diff
@@ -199,13 +199,14 @@ def coarsen_nlvp(problem, self, coefficient_mapping=None):
     if hasattr(problem, "_coarse"):
         return problem._coarse
 
+    def get_coefficients(form):
+        return tuple() if form is None else form.coefficients()
+
     def inject_on_restrict(fine, restriction, rscale, injection, coarse):
         from firedrake.bcs import DirichletBC
         manager = get_transfer_manager(fine)
         finectx = get_appctx(fine)
-        coefficients = finectx.F.coefficients() + finectx.J.coefficients()
-        if finectx.Jp is not None:
-            coefficients += problem.Jp.coefficients()
+        coefficients = chain.from_iterable(map(get_coefficients, (finectx.F, finectx.J, finectx.Jp)))
         coefficients = tuple(OrderedDict.fromkeys(coefficients))
```

Comment:
What does this line even do? Is it needed? At the very least `OrderedDict` isn't needed as Python dicts are now always ordered.

Reply:
Here we get a sorted tuple of unique coefficients, this was present in the first version of this code without no-ops to avoid extra work, and we want to have the same ordering on all ranks because restrict and inject are collective.

Reply:
Got it. Would you mind using [this function](https://github.com/firedrakeproject/firedrake/blob/0fcac11c41ee7b0248d3d9e0fea97d308a6d5edb/firedrake/slate/slate.py#L1392) then? Would be much clearer. The function should also be moved to `utils.py`.

Reply:
Using a `seen = set()` approach is also completely fine. It's just much more readable to have this as a `utils.py` function in my opinion. If the same exists in `variational_solver.py` then that could also be changed.

Reply:
Is a `set()` fine? I thought we were worried about ordering!

Reply:
Yhe `seen = set()` is used differently in variational_solver.py. There we don't want to visit the same functionspace more than once (as opposed to re-visiting the same coefficient that may appear in more than form). The unique approach only makes sense in mg/ufl_utils.py

Reply:
On a second thought, we may first obtain the unique coefficients and then get the unique function spaces to get the unique dms of the problem